kafka-client/src/main/java/dev/responsive/kafka/internal/db/CassandraWindowedTable.java

```diff
@@ -287,15 +314,41 @@ public static CassandraWindowedTable create(
             .build()
     );
 
-    final var setOffset = client.prepare(QueryBuilder
-        .update(name)
-        .setColumn(OFFSET.column(), bindMarker(OFFSET.bind()))
-        .where(PARTITION_KEY.relation().isEqualTo(bindMarker(PARTITION_KEY.bind())))
-        .where(SEGMENT_ID.relation().isEqualTo(bindMarker(SEGMENT_ID.bind())))
-        .where(ROW_TYPE.relation().isEqualTo(METADATA_ROW.literal()))
-        .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(METADATA_KEY)))
-        .where(WINDOW_START.relation().isEqualTo(WINDOW_START.literal(METADATA_TS)))
-        .build()
+    final var setOffset = client.prepare(
+        QueryBuilder
+            .update(name)
+            .setColumn(OFFSET.column(), bindMarker(OFFSET.bind()))
+            .where(PARTITION_KEY.relation().isEqualTo(bindMarker(PARTITION_KEY.bind())))
+            .where(SEGMENT_ID.relation().isEqualTo(bindMarker(SEGMENT_ID.bind())))
+            .where(ROW_TYPE.relation().isEqualTo(METADATA_ROW.literal()))
+            .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(METADATA_KEY)))
+            .where(WINDOW_START.relation().isEqualTo(WINDOW_START.literal(METADATA_TS)))
+            .build()
+    );
+
+    final var fetchStreamTime = client.prepare(
+        QueryBuilder
+            .selectFrom(name)
+            .column(STREAM_TIME.column())
+            .where(PARTITION_KEY.relation().isEqualTo(bindMarker(PARTITION_KEY.bind())))
+            .where(SEGMENT_ID.relation().isEqualTo(bindMarker(SEGMENT_ID.bind())))
+            .where(ROW_TYPE.relation().isEqualTo(METADATA_ROW.literal()))
+            .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(METADATA_KEY)))
+            .where(WINDOW_START.relation().isEqualTo(WINDOW_START.literal(METADATA_TS)))
+            .build()
+    );
+
+    final var setStreamTime = client.prepare(
+        QueryBuilder
+            .update(name)
+            .setColumn(STREAM_TIME.column(), bindMarker(STREAM_TIME.bind()))
+            .where(PARTITION_KEY.relation().isEqualTo(bindMarker(PARTITION_KEY.bind())))
+            .where(SEGMENT_ID.relation().isEqualTo(bindMarker(SEGMENT_ID.bind())))
+            .where(ROW_TYPE.relation().isEqualTo(METADATA_ROW.literal()))
+            .where(DATA_KEY.relation().isEqualTo(DATA_KEY.literal(METADATA_KEY)))
+            .where(WINDOW_START.relation().isEqualTo(WINDOW_START.literal(METADATA_TS)))
+            .ifColumn(EPOCH.column()).isEqualTo(bindMarker(EPOCH.bind()))
```

Comment:
does it make sense here to ensure that the existing stream time <= the desired setStreamTime or is that enforced by the application in all edge cases?

Reply:
I was wondering if adding that condition would have a performance impact since we'd have to read an extra column.

As long as the epoch is checked/we're guaranteed single writer access, I don't see how the stream-time could go backwards, so I don't think it's worth verifying